core/src/components/input-otp/input-otp.tsx

```diff
@@ -0,0 +1,451 @@
+  }
+
+  /**
+   * Get the default value for inputmode based on type if not explicitly set
+   */
+  private getInputmode(): string {
+    const { inputmode } = this;
+    if (inputmode) {
+      return inputmode;
+    }
+
+    if (this.type == 'number') {
+      return 'numeric';
+    } else {
+      return 'text';
+    }
+  }
+
+  private updateValue() {
+    const { inputValues, length } = this;
+    const newValue = inputValues.join('');
+    this.value = newValue;
+    this.ionChange.emit({
+      value: newValue,
+      complete: newValue.length === length,
+    });
+
+    if (newValue.length === length) {
+      this.ionComplete.emit({ value: newValue });
+    }
+  }
+
+  private handleInput(index: number, value: string) {
+    const { validKeys } = this;
+
+    // Only allow input if it's a single character and matches the pattern
+    if (value.length > 1 || (value.length > 0 && !validKeys.test(value.toLowerCase()))) {
+      // Reset the input value if not valid
+      this.inputRefs[index].value = '';
+      this.inputValues[index] = '';
+      return;
+    }
+
+    this.inputValues[index] = value;
+
+    if (value.length > 0) {
+      this.focusNext(index);
+    }
+    this.updateValue();
+  }
+
+  private handleKeyDown(index: number, event: KeyboardEvent) {
+    const { length } = this;
+
+    if (event.key === 'Backspace') {
+      if (this.inputValues[index]) {
+        // Remove the value at the current index
+        this.inputValues[index] = '';
+
+        // Shift all values to the right of the current index left by one
+        for (let i = index; i < length - 1; i++) {
+          this.inputValues[i] = this.inputValues[i + 1];
+        }
+
+        // Clear the last box
+        this.inputValues[length - 1] = '';
+
+        // Update all inputRefs to match inputValues
+        for (let i = 0; i < length; i++) {
+          this.inputRefs[i].value = this.inputValues[i] || '';
+        }
+
+        this.updateValue();
+        event.preventDefault();
+      } else if (!this.inputValues[index] && index > 0) {
+        // If current input is empty, move to previous input
+        this.focusPrevious(index);
+      }
+    } else if (event.key === 'ArrowLeft') {
+      event.preventDefault();
+      this.focusPrevious(index);
+    } else if (event.key === 'ArrowRight') {
+      event.preventDefault();
+      // Only allow moving right if current box has a value
+      if (this.inputValues[index] && index < length - 1) {
+        this.focusNext(index);
+      }
+    } else if (event.key === 'Tab') {
+      // Let all tab events proceed normally
+      return;
+    }
+  }
+
+  private handlePaste(event: ClipboardEvent) {
+    const { inputRefs, length, validKeys } = this;
+
+    event.preventDefault();
+    const pastedText = event.clipboardData?.getData('text') || '';
+    const validChars = pastedText
+      .split('')
+      .filter((char) => validKeys.test(char.toLowerCase()))
```

Comment:
Why using `toLowerCase` here? It seems like `validKeys` considers lower and upper casing via `/^[a-zA-Z0-9]$/i;`, when using a custom pattern, this may be confusing, no?

Reply:
This was a bug, it shouldn't have been case insensitive, fixed: https://github.com/ionic-team/ionic-framework/pull/30386/commits/97649a006b35a7cde236857473961d96c503fa8b